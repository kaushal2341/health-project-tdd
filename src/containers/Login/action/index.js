import {login} from './constants';
export const authenticate = (data) => {
     return {
     payload:data,
     type:login.AUTH
     }
};