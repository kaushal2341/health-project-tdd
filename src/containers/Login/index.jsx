import React from 'react';
import { Form, Message } from 'semantic-ui-react';
import { TextInput, PasswordInput, ButtonInput } from '../../common-ui';
import { postRaw } from '../../axios';
export default class Login extends React.PureComponent {
  state = {
    userCredential: '',
    password: '',
    errorMsg: []
  }
  setTheState = (name, value) => {
    this.setState((prevState) => ({
      [name]: value
    }))
  }

  onChangeHandler = e => {
    const { name, value } = e.target;
    this.setTheState(name, value);
  }

  checkFormSubmitIsValid = () => {
    let errorMsg = [...this.state.errorMsg];
    let formStatus = false;
    Object.keys(this.state).map((s, i) => {
      if (!Object.values(this.state)[i] && s !== 'errorMsg') {
        errorMsg[i] = ` ${s.toUpperCase()} should not be empty`;
        formStatus = false;
      }
      else {
        const patt = /^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}$/;
        const value = Object.values(this.state)[i];
        if (s === 'password' && value.length <= 5) {
          errorMsg[i] = `${s.toUpperCase()} length should be equal or greater than 6`;
          formStatus = false;
        }
        else if (s === 'username' && value.includes('@') && !value.match(patt)) {
          errorMsg[i] = `Not a valid ${s.toUpperCase()}/EMAIL `;
          formStatus = false;
        }
        else {
          errorMsg[i] = '';
          formStatus = true;
        }
      }
    });

    this.setTheState('errorMsg', errorMsg);
    return formStatus;
  }

  apiCall = () => {
    const data = {
      userCredential: this.state.userCredential,
      password: this.state.password
    }
    
    if(this.props.getTokenAfterLogin){
       console.log(data)
       this.props.getTokenAfterLogin(data)
    }   
    // postRaw('/api/login', data).then(response =>{
    //   this.props.submitHandler(response.data);  
    //   console.log('helllo');
    // }).catch(e => {
    //   console.log('error');
    // })  
  }

  submitFormHandler = e => {
    e.preventDefault();
    let isFormValid = this.checkFormSubmitIsValid();
    if (isFormValid)
         this.apiCall();
  }

  render() {
    const depts = [
    { key: '1', text: "Admin", value: "1" },
    { key: '2', text: "Pharmacy", value: "2" },
    { key: '3', text: "Finance", value: "3" }
    ]
    const errorMsg = this.state.errorMsg;
    
    return (
      <>
        <Form>
          <TextInput placeholder="Username/Email" name="userCredential" type="text" onChangeHandler={this.onChangeHandler} value={this.state.userCredential}></TextInput>
          <PasswordInput placeholder="Password" name="password" onChangeHandler={this.onChangeHandler} value={this.state.password}></PasswordInput>
          <ButtonInput id="f-submitter" type='submit' buttonName="Submit" onClickHandler={this.submitFormHandler} ></ButtonInput>
          <Message negative>
            {errorMsg.map((msg, i) => {
              return msg ? <p key={i} >{msg}</p> : ""
            })}
          </Message>
        </Form>
      </>
    )
  }
}

