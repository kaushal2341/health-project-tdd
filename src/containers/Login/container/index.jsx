import React from 'react';
import Login from '../../';
import {connect} from 'react-redux';
import {authenticate} from '../action'
class LoginContainer extends React.PureComponent {
 render(){
     return(
         <Login></Login>
      )
 }
}
const mapStateToProps = state => ({
token:state.token
})
const mapDispatchToProps = dispatch => ({
  login:(data) => authenticate(data)
})

export default connect(mapStateToProps,mapDispatchToProps)(LoginContainer)