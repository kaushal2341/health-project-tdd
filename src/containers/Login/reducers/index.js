import {login} from '../action/constants';        
let state={};

export const loginState =(state,action) =>{
    switch(action.type) {
        case login.AUTH:console.log(state)
                                return ({...state,
                                         token:action.payload
                                         })
        default:return state;
    }
 

};