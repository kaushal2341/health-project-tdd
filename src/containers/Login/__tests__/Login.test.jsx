import React from 'react';
import Login from '..';
import {
  Form,
  Input,
  Button,
  Message
} from 'semantic-ui-react';
import {
  TextInput,
  PasswordInput,
  ButtonInput
} from '../../../common-ui';
expect.addSnapshotSerializer(enzymeSerializer);
describe('Login Container Testing', () => {
  let wrapper, wrapper1;
  beforeEach(() => {
    wrapper = mount(< Login />)
    wrapper1 = shallow(<Login/>)
    
  }); afterEach(() => {
    wrapper.unmount()
  });

  it(' should render the login component successfully ', () => {
    expect(wrapper).toBeDefined();
  });

  it(' should have only one Form tag field ', () => {
    expect(wrapper.find(Form)).toHaveLength(1);
  });

  it('should have only one Text Field ', () => {
    expect(wrapper.find(TextInput)).toHaveLength(1);
  });

  it('should have the give props available to Text Field', () => {
    const textProps = ['placeholder', 'name', 'type', 'onChangeHandler']
    const textInputProps = Object.keys(wrapper1.find(TextInput).props());
    textProps.map((textProp, i) => {
      expect(textProps).toContain(textInputProps[i]);
    });
  });

  it('should have onChangeHandler function', () => {
    expect(wrapper.instance().onChangeHandler).toBeDefined();
  });

  it('should have only one Password Field', () => {
    expect(wrapper.find(PasswordInput)).toHaveLength(1);
  });

  it('should have the given props available for password field', () => {
    const passProps = ['placeholder', 'name', 'onChangeHandler']
    const passInputProps = Object.keys(wrapper1.find(PasswordInput).props())
    passProps.map((passProp, i) => {
      expect(passProps).toContain(passInputProps[i]);
    });
  });

  it('should have only one submit Button Field', () => {
    expect(wrapper1.find(ButtonInput)).toHaveLength(1);
  });

  it('should have the given props available for button', () => {
    const buttonProps = ['type', 'buttonName', 'onClickHandler']
    const buttonInputProps = Object.keys(wrapper1.find(ButtonInput).props())
    buttonProps.map((buttonProp, i) => {
      expect(buttonProps).toContain(buttonInputProps[i]);
    });
  });

  it('should have submitFormHandler function', () => {
    expect(wrapper.instance().submitFormHandler).toBeDefined();
  })

  it('should have state for all the input field except button', () => {
    const states = ['userCredential', 'password', 'errorMsg'];
    const statesFromComp = Object.keys(wrapper.state());
    states.map((state, i) => {
      expect(states).toContain(statesFromComp[i]);
    })
  });

  it('should have atleast one Message field', () => {
    expect(wrapper.find(Message)).toHaveLength(1);
  })

  describe('Input Simulation Change', () => {
    it('should change the value of state of username when Text Input is changed', () => {
      wrapper.find(TextInput).find(Input).find('input').simulate('change', {
        target: {
          name: "username",
          value: "kaushal@gmail.com"
        }
      });
      expect(wrapper.state('username')).toEqual('kaushal@gmail.com');
    });

    it('should change the value of state of password when Password Input is changed', () => {
      wrapper.find(PasswordInput).find(Input).find('input').simulate('change', {
        target: {
          name: "password",
          value: "Flasher234@@"
        }
      });
      expect(wrapper.state('password')).toEqual('Flasher234@@');
    });

    describe('Form Submit Simulation', () => {
      let submitCallBack;
      beforeEach(() => {
        submitCallBack = jest.spyOn(wrapper.instance(), 'submitFormHandler');
        wrapper.instance().forceUpdate();

      });
      it('should submit the form and prevent reloading when form is submitted with non empty value', () => {
        wrapper.find(ButtonInput).last().find(Button).find('button').simulate('click')
        expect(submitCallBack).toHaveBeenCalled()
      });

      it('should not submit the form if all the value is not provided and provide errorMsg for form', () => {

        wrapper.find(ButtonInput).last().find(Button).find('button').simulate('click')
        expect(wrapper.state('errorMsg').length).toBeGreaterThan(0);
      });

      it('should submit the form if all the value is  provided and donot provide errorMsg for form', () => {
        wrapper.setState({
          username: 'kaushal123',
          password: 'flasher@123'
        })
        expect(wrapper.state('errorMsg').length).toBe(0)
      })
    });

});


describe('Snap Shot Testing', () => {
    it('should create or match previous snapshot of the component', () => {
      expect(wrapper1).toMatchSnapshot();
    });
  });
});