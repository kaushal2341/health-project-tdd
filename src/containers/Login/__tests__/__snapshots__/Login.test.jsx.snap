// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Container Testing Snap Shot Testing should create or match previous snapshot of the component 1`] = `
<Login>
  <Form
    as="form"
  >
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <TextInput
        name="username"
        onChangeHandler={[Function]}
        placeholder="Username/Email"
        type="text"
        value=""
      >
        <Input
          name="username"
          onChange={[Function]}
          placeholder="Username/Email"
          type="text"
          value=""
        >
          <div
            className="ui input"
          >
            <input
              name="username"
              onChange={[Function]}
              placeholder="Username/Email"
              type="text"
              value=""
            />
          </div>
        </Input>
        <span
          id="error"
        />
      </TextInput>
      <PasswordInput
        name="password"
        onChangeHandler={[Function]}
        placeholder="Password"
        value=""
      >
        <Input
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        >
          <div
            className="ui input"
          >
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
        </Input>
        Â 
        <ButtonInput
          id="btn_error"
          onClickHandler={[Function]}
          size="tiny"
        >
          <Button
            as="button"
            id="btn_error"
            onClick={[Function]}
            size="tiny"
          >
            <Ref
              innerRef={
                Object {
                  "current": <button
                    class="ui tiny button"
                    id="btn_error"
                  >
                    <i
                      aria-hidden="true"
                      class="icon eye slash"
                    />
                  </button>,
                }
              }
            >
              <RefFindNode
                innerRef={
                  Object {
                    "current": <button
                      class="ui tiny button"
                      id="btn_error"
                    >
                      <i
                        aria-hidden="true"
                        class="icon eye slash"
                      />
                    </button>,
                  }
                }
              >
                <button
                  className="ui tiny button"
                  id="btn_error"
                  onClick={[Function]}
                >
                  <Icon
                    as="i"
                    className="eye slash"
                  >
                    <i
                      aria-hidden="true"
                      className="icon eye slash"
                      onClick={[Function]}
                    />
                  </Icon>
                </button>
              </RefFindNode>
            </Ref>
          </Button>
        </ButtonInput>
        <span
          id="error"
        />
      </PasswordInput>
      <ButtonInput
        buttonName="Submit"
        onClickHandler={[Function]}
        type="submit"
      >
        <Button
          as="button"
          onClick={[Function]}
          type="submit"
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button"
                  type="submit"
                >
                  Submit
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button"
                    type="submit"
                  >
                    Submit
                  </button>,
                }
              }
            >
              <button
                className="ui button"
                onClick={[Function]}
                type="submit"
              >
                Submit
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </ButtonInput>
      <Message
        negative={true}
      >
        <div
          className="ui negative message"
        />
      </Message>
    </form>
  </Form>
</Login>
`;
